content/blog/gotchaR-bumps.md

```diff
@@ -0,0 +1,162 @@
+though I don't think I should need it, I instinctively put `data$thing <- NULL` as the 
+second line. It makes no difference.
+
+I restart RStudio, reboot my computer, update R - the usual sort of random flailing thing. 
+All useless. I start to wonder if R has been tricking me all along, and I do some really
+basic assignments and concatenations to check it's not been fatally flawed all these years. 
+It appears fundamentally sound. 
+
+Eventually, I ask my more R-experienced colleagues if, in fact, I am going insane. They
+cannot answer this for sure, but looking at the programming issue, they tell me it is a 
+feature of R in a way. 
+
+The first time I read from `data$thing`, it doesn't exist 
+in the list, and R lexically auto-completes it into `data$thing_type` - hence the 
+"Fruit". It doesn't auto-complete on the assignment side though, so after the first
+loop iteration, `data$thing` exists, and things continue normally. 
+
+It's tempting to think
+`data$thing <- NULL` beforehand would help - but with a list, that
+would only remove the (non-existent) `thing` from list, because that's what NULL assignment
+with lists does.
+
+So what should I instead do? One option is to replace my inner loop with
+` data$thing <- c(data[['thing']], f)`, because the auto-completion is activated on the 
+use of the dollar sign, but not on a quoted variable name. Moreover, setting these options:
+```
+warnPartialMatchAttr = TRUE
+warnPartialMatchDollar = TRUE
+warnPartialMatchArgs = TRUE
+``` in `.Rprofile` will throw a warning every time R does a partial-match. This seems like
+it might save a world of pain, although casting my mind back to various projects where I have
+frivolously talked about `list$thing` and `list$thing_name` and `list$thing_id` still makes me
+shudder a little. 
+
+Also having set these options, I occasionally see warnings popping up about auto-completion
+in code for other libraries we rely upon...
+
+## data.table vs data.frame
+
+So I collected some more data for my colleagues, and I put it in a `data.frame` like this.
+```
+data <- data.frame(stringsAsFactors = FALSE,
+  name = c("Alex", "Emma", "James", "Rich", "Rob"),
+  food = c("Oysters", "Haggis", "Goose", "Cheese", "Celery"))
+
+who_likes <- function(food) {
+  data[data$food == food, ]
+}
+```
+
+This is very nice. So if I have a haggis and wonder what to do with it:
+```
+> who_likes("Haggis")
+  name   food
+2 Emma Haggis
+```
+
+Great. However, I later want to do something astoundingly complicated with this rich dataset, and I 
+decide that using the popular [data.table](https://rdatatable.gitlab.io/data.table/) will 
+be a good idea, since it's just a higher performance version of `data.frame`, right? It's
+also in very common use. (To be clear, the performance increase is indeed very worthwhile for large
+datasets).
+
+However... if I swap in the `data.table` by only changing 
+```
+data <- data.frame(stringsAsFactors = FALSE
```

Comment:
missing closing paren here and line 113 - consider using `...` as `data.frame(..., stringsAsFactors = FALSE)` perhaps?

Reply:
I've cleaned that up a bit into a tighter sentence - it was all a bit spaced out anyway.